iroh/src/magicsock.rs

```diff
@@ -1561,7 +1646,67 @@ enum DiscoBoxError {
     Parse(anyhow::Error),
 }
 
-type RelayRecvResult = Result<(PublicKey, quinn_udp::RecvMeta, Bytes), io::Error>;
+/// Channel for [`MagicSock::poll_recv_relay`] to receive datagrams from relays.
+///
+/// The sender and receiver will take care of the required wakers needed for
+/// [`AsyncUdpSocket::poll_recv`].
+// TODO: This channel should possibly be implemented with concurrent-queue and atomic-waker.
+// Or maybe async-channel.
+fn relay_recv_channel() -> (RelayRecvSender, RelayRecvReceiver) {
+    let (tx, rx) = mpsc::channel(128);
+    let waker = Arc::new(parking_lot::Mutex::new(None));
+    let sender = RelayRecvSender {
+        sender: tx,
+        waker: waker.clone(),
+    };
+    let receiver = RelayRecvReceiver {
+        receiver: parking_lot::Mutex::new(rx),
+        waker,
+    };
+    (sender, receiver)
+}
+
+#[derive(Debug, Clone)]
+struct RelayRecvSender {
+    sender: mpsc::Sender<RelayRecvDatagram>,
+    waker: Arc<parking_lot::Mutex<Option<Waker>>>,
+}
+
+impl RelayRecvSender {
+    fn try_send(
+        &self,
+        item: RelayRecvDatagram,
+    ) -> Result<(), mpsc::error::TrySendError<RelayRecvDatagram>> {
+        self.sender.try_send(item).inspect(|_| {
+            if let Some(waker) = self.waker.lock().take() {
+                waker.wake();
+            }
+        })
+    }
+}
+
+#[derive(Debug)]
+struct RelayRecvReceiver {
+    receiver: parking_lot::Mutex<mpsc::Receiver<RelayRecvDatagram>>,
+    waker: Arc<parking_lot::Mutex<Option<Waker>>>,
+}
+
+impl RelayRecvReceiver {
+    fn poll_recv(&self, cx: &mut Context) -> Poll<Result<RelayRecvDatagram>> {
+        let mut receiver = self.receiver.lock();
+        self.waker.lock().replace(cx.waker().clone());
+        match receiver.try_recv() {
+            Ok(item) => {
+                self.waker.lock().take();
+                Poll::Ready(Ok(item))
+            }
+            Err(mpsc::error::TryRecvError::Empty) => Poll::Pending,
+            Err(mpsc::error::TryRecvError::Disconnected) => {
+                Poll::Ready(Err(anyhow!("All RelayRecvSenders disconnected")))
+            }
+        }
+    }
```

Comment:
Could folks, especially @matheus23 as he spotted this, review this again?  The previous fix for this was horrendously broken.  This I think is more correct.

Reply:
You're totally right. There was a potential for a race but it looks different than what I thought :upside_down_face: 

Let's not waste a lot of time on this and instead just fix it by getting rid of it in a follow-up that replaces this with an async-channel or concurrent queue.

Reply:
I wonder why you think the previous logic was horrendously broken. Did it fail tests?

Reply:
A broken test run: https://github.com/n0-computer/iroh/actions/runs/12143158923/job/33859383227?pr=2986